File: julia/words/test/WordsTest.jl
#!/usr/bin/julia

using Test
include("../src/Words.jl")
using .Words

@testset "Words Suite" begin
  @testset "Reverse String" begin
    @test reverse_string("hello") == "olleh"
    @test reverse_string("a") == "a"
    @test reverse_string("") == ""
  end

  @testset "Is Palindrome" begin
    @test is_palindrome("race car") == true
    @test is_palindrome("level") == true
    @test is_palindrome("hello") == false
    @test is_palindrome("a") == true
    @test is_palindrome("") == true
  end

  @testset "Is Substring" begin
    @test kmp_search("test", "this is a test") == true
    @test kmp_search("not", "this is a test") == false
    @test kmp_search("", "any string") == true
    @test kmp_search("abc", "abc") == true
    @test kmp_search("abc", "ab") == false
  end

  @testset "Longest Common Subsequence" begin
    @test lcs("AGGTAB", "GXTXAYB") == 4
    @test lcs("ABC", "ABC") == 3
    @test lcs("ABC", "DEF") == 0
    @test lcs("", "ABC") == 0
    @test lcs("ABC", "") == 0
  end
end

File: scala/words/src/test/scala/words_test.scala
import org.scalatest.funsuite.AnyFunSuite

class WordsTest extends AnyFunSuite {
	test("reverse string") {
		assert(Words.reverseString("hello") == "olleh")
		assert(Words.reverseString("a") == "a")
		assert(Words.reverseString("") == "")
	}

	test("is palindrome") {
		assert(Words.isPalindrome("racecar") == true)
		assert(Words.isPalindrome("level") == true)
		assert(Words.isPalindrome("hello") == false)
		assert(Words.isPalindrome("a") == true)
		assert(Words.isPalindrome("") == true)
	}

	test("is substring") {
		assert(Words.kmpSearch("test", "this is a test") == true)
		assert(Words.kmpSearch("not", "this is a test") == false)
		assert(Words.kmpSearch("", "any string") == true)
		assert(Words.kmpSearch("abc", "abc") == true)
		assert(Words.kmpSearch("abc", "ab") == false)
	}

	test("longest common subsequence") {
		assert(Words.lcs("AGGTAB", "GXTXAYB") == 4)
		assert(Words.lcs("ABC", "ABC") == 3)
		assert(Words.lcs("ABC", "DEF") == 0)
		assert(Words.lcs("", "ABC") == 0)
		assert(Words.lcs("ABC", "") == 0)
	}
}
		